src/daft-core/src/array/ops/cast.rs

```diff
@@ -1527,6 +1601,86 @@ impl TensorArray {
     }
 }
 
+impl COOSparseTensorArray {
+    pub fn cast(&self, dtype: &DataType) -> DaftResult<Series> {
+        match dtype {
+            DataType::Tensor(inner_dtype) => {
+                let non_zero_values_array = self.values_array();
+                let non_zero_indices_array = self.indices_array();
+                let shape_array = self.shape_array();
+                let mut sizes_vec: Vec<usize> = vec![0; shape_array.len()];
+                for (i, shape) in shape_array.into_iter().enumerate() {
+                    match shape {
+                        Some(shape) => {
+                            let shape = shape.u64().unwrap().as_arrow();
+                            let num_elements =
+                                shape.values().clone().into_iter().product::<u64>() as usize;
+                            sizes_vec[i] = num_elements;
+                        }
+                        _ => {}
+                    }
+                }
+                let offsets: Offsets<i64> = Offsets::try_from_iter(sizes_vec.iter().cloned())?;
+                let mut values = vec![0 as i64; sizes_vec.iter().sum::<usize>() as usize];
+                for (i, indices) in non_zero_indices_array.into_iter().enumerate() {
+                    for j in 0..indices.unwrap().len() {
+                        let index = non_zero_indices_array
+                            .get(i)
+                            .unwrap()
+                            .u64()
+                            .unwrap()
+                            .as_arrow()
+                            .value(j) as usize;
+                        let list_start_offset = offsets.start_end(i).0;
+                        values[list_start_offset + index] = non_zero_values_array
+                            .get(i)
+                            .unwrap()
+                            .i64()
+                            .unwrap()
+                            .as_arrow()
+                            .value(j);
+                    }
```

Comment:
Here I assume the type of the values not sure how to both be able to edit the physical memory and keep it dynamically typed.

Reply:
Not sure it is the idiomatic way to do it but, you can create a macro that accepts the inner_dtype and the primitive type it is mapped to, initialize the vector with the primitive and downcast the series with the array type, for example:

```rust
macro_rules! implement_cast_for_dense_with_inner_dtype {
    ($dtype:ty, $array_type:ty, $n_values:ident, $non_zero_indices_array:ident, $non_zero_values_array:ident, $offsets:ident) => {{
        let mut values = vec![0 as $dtype; $n_values];
        for (i, indices) in $non_zero_indices_array.into_iter().enumerate() {
            for j in 0..indices.unwrap().len() {
                let index = $non_zero_indices_array
                    .get(i)
                    .unwrap()
                    .u64()
                    .unwrap()
                    .as_arrow()
                    .value(j) as usize;
                let list_start_offset = $offsets.start_end(i).0;
                values[list_start_offset + index] = $non_zero_values_array
                    .get(i)
                    .unwrap()
                    .downcast::<$array_type>()
                    .unwrap()
                    .as_arrow()
                    .value(j);
            }
        }
        Box::new(arrow2::array::PrimitiveArray::from_vec(values))
    }};
}

impl COOSparseTensorArray {
    pub fn cast(&self, dtype: &DataType) -> DaftResult<Series> {
        match dtype {
            DataType::Tensor(inner_dtype) => {
                let non_zero_values_array = self.values_array();
                let non_zero_indices_array = self.indices_array();
                let shape_array = self.shape_array();
                let mut sizes_vec: Vec<usize> = vec![0; shape_array.len()];
                for (i, shape) in shape_array.into_iter().enumerate() {
                    match shape {
                        Some(shape) => {
                            let shape = shape.u64().unwrap().as_arrow();
                            let num_elements =
                                shape.values().clone().into_iter().product::<u64>() as usize;
                            sizes_vec[i] = num_elements;
                        }
                        _ => {}
                    }
                }
                let offsets: Offsets<i64> = Offsets::try_from_iter(sizes_vec.iter().cloned())?;
                let n_values = sizes_vec.iter().sum::<usize>() as usize;
                let item: Box<dyn arrow2::array::Array> = match inner_dtype.as_ref() {
                    DataType::Float32 => implement_cast_for_dense_with_inner_dtype!(f32, Float32Array, n_values, non_zero_indices_array, non_zero_values_array, offsets),
                    DataType::Int64 => implement_cast_for_dense_with_inner_dtype!(i64, Int64Array, n_values, non_zero_indices_array, non_zero_values_array, offsets),
                    _ => panic!("Hi")
                };
                let list_arr = ListArray::new(
                    Field::new(
                        "data",
                        DataType::List(Box::new(inner_dtype.as_ref().clone())),
                    ),
                    Series::try_from((
                        "item",
                        item,
                    ))?,
                    offsets.into(),
                    None,
                ).into_series();
                let physical_type = dtype.to_physical();
                let struct_array = StructArray::new(
                    Field::new(self.name(), physical_type),
                    vec![list_arr, shape_array.clone().into_series()],
                    None
                );
                Ok(
                    TensorArray::new(Field::new(self.name(), dtype.clone()), struct_array)
                        .into_series(),
                )
            }
            (_) => self.physical.cast(dtype),
        }
    }
}
```